pkg/autoscaler/metrics/stats_scraper.go

```diff
@@ -132,7 +137,60 @@ func urlFromTarget(t, ns string) string {
 		t, ns, networking.AutoscalingQueueMetricsPort)
 }
 
-// Scrape calls the destination service then sends it
+// ScrapeForRemoval collects metrics from all pods for decision making on removal
```

Comment:
IIRC this isn't what was here. Scraping **all** pods without current scraping strategy is also going to cause big-time overhead for larger deployments. Not necessarily due to the number of pods to scrape but because we scrape through the service and need to retry if we hit the same pod twice. Capturing 100% of all pods in say a 100 pod deployment might take... a while.

Reply:
I agree that scraping is a point of concern here. But async scraping can result in suboptimal scaling decisions too, depending on what pods we scrape. 

How about the following:
- short term solution, have a cap on the max number of pods we scrape before falling back on sampling (let's say 10 or 20?)
- long term, use the per node DaemonSet scraper to gather data on all pods

Reply:
So here's my take on it. It's all about numbers...

We already do sampling and for numbers < 5, I think sample size == pod count. At 10 we sample 7 (my memory might be shaky, but we have a UT with numbers). 

So if your revision is small or not big we're already sampling most of the pods. Which, while not ideal _scale down_ still provides a better state than what we have now. And, fwiw, it probably would pass the test that I am yearning to see -- a real life benchmark of this actuating real life revisions with cc > 1. 

If your revision is big, then you just have to sample, otherwise you're blocking the pipeline of autoscaler and we know there are changes happening right now, since otherwise this would not be triggered in the first place right? So with larger revisions — we're sampling anyway. Think hundreds of pods and one is terminating, since one request dropped of (cc=1, or actually any value thereof, if it's next to the threshold). Which one are you going to kill?

Now, what would help here is a definition in probabilistic terms: _how good do we want the scale down to be_.
Is it 95% of pods that will scale down are in the set of minimally used pods?
Do we need to tune the sampling algorithm to yield a different curve for this, than for sampling?



_at this point it feels like I'm rambling, so let me know if this does not make sense_ 😭

Reply:
I had a slack conversation with @vagababov on some of the points above.

We do `sample size == pod count` if `pods <= 3`, otherwise we sample to achieve 95% probabilistic confidence. I am just wondering if `3` is too small of a number. For `TestGracefulScaledown` e2e to consistently pass, we need sample size to be minimum `4` but ideally even slightly larger in case more pods get created in response to the panic mode caused by the test. But in a larger context, outside the test, maybe we can afford to have it be a bigger number, in case people turn the feature on.

I however think the DaemonSet scraping can be vastly useful here. The biggest problem we have with current sampling strategy is the false positive we get when hitting the metrics service. With a DaemonSet scraper, there wont be any false positive. Even with a clsuter of 50 nodes, we can scrape all 50 nodes, in parallel, without any false positive. This will hopefully get us over the hump.

Finally on the e2e test, the test is flaking because the autoscaler still exhibits some weird behavior when creating early count of pods. I see it scaling up to a given number of pods, terminating some, and starting some other. This confuses the graceful scale down e2e on what pods it should expect to have open connections to.

Reply:
Never forget pod addressability issues or rather the lack thereof.

This is a distributed system. We need to make peace with our decisions needing to be based on imperfect information. Even if you scrape every single pod, there's no guarantees that the pods you scraped didn't get a connection afterwards.

Reply:
yes, pod addressability :sigh: ....

```
nimak  5:17 PM
yeah. thats why I think no mesh + daemon set is reasonable
30 nodes scraped in parallel isnt too much overhead
its hit every time … (until mesh is enabled :smile: ) 

vagababov  5:18 PM
Yeah maybe
Need more measurements and numbers
```